Core/include/Acts/Propagator/Propagator.ipp

```diff
@@ -27,80 +30,154 @@ template <typename S, typename N>
+    }
+    // Accumulate the path length
+    state.pathLength += *res;
+    // Update the position and direction
+    state.position = m_stepper.position(state.stepping);
+    state.direction =
+        state.options.direction * m_stepper.direction(state.stepping);
+
+    ACTS_VERBOSE("Step with size " << *res << " performed. We are now at "
+                                   << state.position.transpose()
+                                   << " with direction "
+                                   << state.direction.transpose());
+
+    // release actor and aborter constrains after step was performed
+    m_stepper.releaseStepSize(state.stepping, ConstrainedStep::navigator);
+    m_stepper.releaseStepSize(state.stepping, ConstrainedStep::actor);
+
+    // Post-stepping: check target status, call actors, check abort conditions
+    state.stage = PropagatorStage::postStep;
+
+    if (nextTarget.isValid()) {
+      IntersectionStatus postStepSurfaceStatus = m_stepper.updateSurfaceStatus(
+          state.stepping, *nextTarget.surface,
+          nextTarget.surfaceIntersectionIndex, state.options.direction,
+          nextTarget.boundaryTolerance, state.options.surfaceTolerance,
+          logger());
+      if (postStepSurfaceStatus == IntersectionStatus::onSurface) {
+        m_navigator.handleSurfaceReached(state.navigation, state.position,
+                                         state.direction, *nextTarget.surface);
+      }
+      if (postStepSurfaceStatus != IntersectionStatus::reachable) {
+        nextTarget = NavigationTarget::invalid();
+      }
+    }
 
-  state.stage = PropagatorStage::postPropagation;
+    state.options.actorList.act(state, m_stepper, m_navigator, logger());
 
-  // if we didn't terminate normally (via aborters) set navigation break.
-  // this will trigger error output in the lines below
+    if (state.options.actorList.checkAbort(state, m_stepper, m_navigator,
+                                           logger())) {
+      terminatedNormally = true;
+      break;
+    }
+
+    // Update the position and direction because actors might have changed it
+    state.position = m_stepper.position(state.stepping);
+    state.direction =
+        state.options.direction * m_stepper.direction(state.stepping);
+
+    // Pre-Stepping: target setting
+    state.stage = PropagatorStage::preStep;
+
+    if (nextTarget.isValid() &&
+        !m_navigator.checkTargetValid(state.navigation, state.position,
+                                      state.direction)) {
+      ACTS_VERBOSE("Target is not valid anymore.");
+      nextTarget = NavigationTarget::invalid();
+    }
+
+    if (!nextTarget.isValid()) {
+      // navigator step constraint is not valid anymore
+      m_stepper.releaseStepSize(state.stepping, ConstrainedStep::navigator);
+
+      nextTargetResult = getNextTarget();
+      if (!nextTargetResult.ok()) {
+        return nextTargetResult.error();
+      }
+      nextTarget = *nextTargetResult;
```

Comment:
If this is invalid, what's the path we take? I assume this eventually leads to the normal termination of the propagation / navigation but it's not clear to me how that happens.

Reply:
Discussed offline: we tried to improve the flow a bit by disambiguating `invalid` and `none` target